pkg/controller/oam/applicationconfiguration/apply.go

```diff
@@ -19,61 +19,210 @@ package applicationconfiguration
+			}
+		}
+	}
+
+	return nil
+}
+
+func findDereferencedScopes(statusScopes []v1alpha2.WorkloadScope, scopes []unstructured.Unstructured) []v1alpha2.WorkloadScope {
+	toBeDeferenced := []v1alpha2.WorkloadScope{}
+	for _, ss := range statusScopes {
+		found := false
+		for _, s := range scopes {
+			if (s.GetAPIVersion() == ss.Reference.APIVersion) &&
+				(s.GetKind() == ss.Reference.Kind) &&
+				(s.GetName() == ss.Reference.Name) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			toBeDeferenced = append(toBeDeferenced, ss)
+		}
+	}
+
+	return toBeDeferenced
+}
+
+func (a *workloads) applyScope(ctx context.Context, wl Workload, s unstructured.Unstructured) error {
+	workloadRef := runtimev1alpha1.TypedReference{
+		APIVersion: wl.Workload.GetAPIVersion(),
+		Kind:       wl.Workload.GetKind(),
+		Name:       wl.Workload.GetName(),
+	}
+
+	// Get Scope again to make sure we work with the most up-to date object.
+	scopeObject := unstructured.Unstructured{}
+	scopeObject.SetAPIVersion(s.GetAPIVersion())
+	scopeObject.SetKind(s.GetKind())
+	scopeObjectRef := types.NamespacedName{Namespace: s.GetNamespace(), Name: s.GetName()}
+	if err := a.rawClient.Get(ctx, scopeObjectRef, &scopeObject); err != nil {
+		return errors.Wrapf(err, errFmtApplyScope, s.GetName())
+	}
+
+	refs := []interface{}{}
+	// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+	if value, err := fieldpath.Pave(scopeObject.UnstructuredContent()).GetValue("status.workloadRefs"); err == nil {
+		refs = value.([]interface{})
+
+		for _, item := range refs {
+			ref := item.(map[string]interface{})
+			if (workloadRef.APIVersion == ref["apiVersion"]) &&
+				(workloadRef.Kind == ref["kind"]) &&
+				(workloadRef.Name == ref["name"]) {
+				return nil
+			}
+		}
+	}
+
+	refs = append(refs, workloadRef)
+	// TODO(asouza): Need to find a way to make sure that this is compatible with the scope structure.
+	if err := fieldpath.Pave(scopeObject.UnstructuredContent()).SetValue("status.workloadRefs", refs); err != nil {
+		return errors.Wrapf(err, errFmtSetWorkloadRef, s.GetName(), wl.Workload.GetName())
+	}
+
+	if err := a.rawClient.Status().Update(ctx, &scopeObject); err != nil {
+		return errors.Wrapf(err, errFmtApplyScope, s.GetName())
+	}
+
+	return nil
+}
+
+func (a *workloads) applyScopeRemoval(ctx context.Context, namespace string, ws v1alpha2.WorkloadStatus, s v1alpha2.WorkloadScope) error {
+	workloadRef := runtimev1alpha1.TypedReference{
+		APIVersion: ws.Reference.APIVersion,
+		Kind:       ws.Reference.Kind,
+		Name:       ws.Reference.Name,
+	}
+
+	// Get Scope again to make sure we work with the most up-to date object.
+	scopeObject := unstructured.Unstructured{}
+	scopeObject.SetAPIVersion(s.Reference.APIVersion)
+	scopeObject.SetKind(s.Reference.Kind)
+	scopeObjectRef := types.NamespacedName{Namespace: namespace, Name: s.Reference.Name}
+	if err := a.rawClient.Get(ctx, scopeObjectRef, &scopeObject); err != nil {
```

Comment:
I am not sure if we need to fetch so many times in the code, it won't guarantee anything anyway.

Reply:
why fetch instead of use the object?

Reply:
better not to fetch again to avoid inconsistency